pkg/derivation/kv.go

```diff
@@ -0,0 +1,94 @@
+package derivation
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"os"
+)
+
+type KVStore interface {
+	Get(key string) ([]byte, error)
+	Set(key string, value []byte) error
+}
+
+// File system KV
+
+type FSKVStore struct{}
+
+func (kv *FSKVStore) Get(key string) ([]byte, error) {
+	return os.ReadFile(key)
+}
+
+func (kv *FSKVStore) Set(key string, value []byte) error {
+	return fmt.Errorf("writing not supported")
+}
+
+// Input substitution calculations
+
+type InputSubstKV struct {
+	memo  map[string][]byte // Memoized values
+	store KVStore           // Underlying store
+}
+
+func NewInputSubstKV(store KVStore) *InputSubstKV {
+	return &InputSubstKV{
+		memo:  make(map[string][]byte),
+		store: store,
+	}
+}
+
+func (kv *InputSubstKV) Get(key string) ([]byte, error) {
```

Comment:
This doesn't seem to return what you passed in, but does a whole bunch of calculation on what it from the underlying KV.

It should probably be named differently - and I'd argue we want to do that calculation once, while calling Set(), and then just return the cached result.